src/coreclr/src/jit/morph.cpp

```diff
@@ -18370,6 +18382,106 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
+        }
+
+        LclVarDsc* varDsc = m_compiler->lvaGetDesc(val.LclNum());
+
+        if (varDsc->TypeGet() != TYP_STRUCT)
+        {
+            // TODO-ADDR: Skip integral/floating point variables for now, they're more
+            // complicated to transform. We can always turn an indirect access of such
+            // a variable into a LCL_FLD but that blocks enregistration so we need to
+            // detect those case where we can use LCL_VAR instead, perhaps in conjuction
+            // with CAST and/or BITCAST.
+            // Also skip SIMD variables for now, fgMorphFieldAssignToSIMDIntrinsicSet and
+            // others need to be updated to recognize LCL_FLDs.
+            return;
+        }
+
+        if (varDsc->lvPromoted || varDsc->lvIsStructField || m_compiler->lvaIsImplicitByRefLocal(val.LclNum()))
+        {
+            // TODO-ADDR: For now we ignore promoted and "implict by ref" variables,
+            // they require additional changes in subsequent phases
+            // (e.g. fgMorphImplicitByRefArgs does not handle LCL_FLD nodes).
+            return;
+        }
+
+        if (val.Offset() > UINT16_MAX)
+        {
+            // TODO-ADDR: We should morph the address of the indir using MorphLocalAddress,
+            // once the current usage restrictions on LCL_VAR|FLD_ADDR are lifted.
+            return;
+        }
+
+        FieldSeqNode* fieldSeq = val.FieldSeq();
+
+        if (indir->OperIs(GT_IND) && (fieldSeq != nullptr) && (fieldSeq != FieldSeqStore::NotAField()))
+        {
+            // If we have an IND node and a field sequence then they should have the same type.
+            // Otherwise it's best to forget the field sequence since the resulting LCL_FLD
+            // doesn't match a real struct field. Value numbering to protect itself from such
+            // mismatches but there doesn't seem to be any good reason to generate a LCL_FLD
+            // with a mismatched field sequence only to have to ignore it later.
+
+            assert(!fieldSeq->IsPseudoField());
+
+            if (indir->TypeGet() != JITtype2varType(m_compiler->info.compCompHnd->getFieldType(fieldSeq->m_fieldHnd)))
+            {
+                fieldSeq = nullptr;
+            }
+        }
+
+        indir->ChangeOper(GT_LCL_FLD);
+        indir->AsLclFld()->SetLclNum(val.LclNum());
+        indir->AsLclFld()->SetLclOffs(val.Offset());
+        indir->AsLclFld()->SetFieldSeq(fieldSeq == nullptr ? FieldSeqStore::NotAField() : fieldSeq);
+
+        unsigned flags = 0;
+
+        if (user->OperIs(GT_ASG) && (user->AsOp()->gtGetOp1() == indir))
+        {
+            flags |= GTF_VAR_DEF | GTF_DONT_CSE;
+
+            if (genTypeSize(indir->TypeGet()) < m_compiler->lvaLclExactSize(val.LclNum()))
+            {
+                flags |= GTF_VAR_USEASG;
+            }
+        }
+
+        indir->gtFlags = flags;
+
+        m_compiler->lvaSetVarDoNotEnregister(val.LclNum() DEBUGARG(Compiler::DNER_LocalField));
```

Comment:
It might be worth a comment here to say that we've already done struct promotion, so if we're accessing a field that hasn't been promoted, the struct cannot be enregistered.

Reply:
Yes, and currently we also skip all this code if we're dealing with a promoted struct var.